# **EDA and Data Cleaning (Dealing with Missing Values)**

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

**Identifying NaNs are very important because they have a `carry over` effect:**

In [4]:
a = 2
b = np.nan

print(f'The result of {a} and {b} added is {a + b}')

The result of 2 and nan added is nan


This is means if you add any number to a NaN the result is a NaN.

**Likewise it happens with multiplication:**

In [7]:
a = 2
b = np.nan

print(f'The result of {a} and {b} mutiplied is {a * b}')

The result of 2 and nan mutiplied is nan


- These NaNs affects all of the caluclations like averages and standard deviations. 

## **Exploratory Data Analysis (EDA):**

To conduct a full Exploratory Data Analysis (EDA) on the dataset, we will start by loading the data to understand its structure, content, and the types of data it contains. This will involve:

- Checking the first few rows to get a feel of the data.
- Understanding the data types of each column and look for any initial issues with the data.
- Checking for missing values.
- Summarising the statistical properties of the numerical data.
- Exploring the distribution of categorical variables.
- Visualising the data to uncover trends and patterns.

### **First we load the dataset and view the first few rows**:

In [12]:
# Load the dataset
drinks_data = pd.read_csv('drinks.csv')

# Show the first 5 columns of the dataset
drinks_data.head()

country  beer_servings  spirit_servings  wine_servings  \
0  Afghanistan              0                0              0   
1      Albania             89              132             54   
2      Algeria             25                0             14   
3      Andorra            245              138            312   
4       Angola            217               57             45   

   total_litres_of_pure_alcohol continent  
0                           0.0        AS  
1                           4.9        EU  
2                           0.7        AF  
3                          12.4        EU  
4                           5.9        AF

The dataset contains the following columns:
- **Country** - The country where the drinks are from
- **Beer_servings** - The amount of drinks served
- **Spirit_servings** - The amount of spirits served
- **Wine_servings** - The amount of wine served
- **Total_litres_of_pure_alcohol** - The total litres of alcohol the was served in the respective countries
- **Continent** - The continent of which the drinks are being sold

### **Next, Understanding the data types of each columns and looking for any issues with the data:**

In [15]:
# Check the datatypes of each column
data_types = drinks_data.dtypes

data_types

country                          object
beer_servings                     int64
spirit_servings                   int64
wine_servings                     int64
total_litres_of_pure_alcohol    float64
continent                        object
dtype: object

### **Checking for missing values in the Data:**

In [17]:
# Checking for missing values in columns
drinks_data.isna().sum()

country                          0
beer_servings                    0
spirit_servings                  0
wine_servings                    0
total_litres_of_pure_alcohol     0
continent                       23
dtype: int64

### **Summarising the statistial properties of the numerical data:**

In [19]:
# Summary statistics for numerical data
summary_stats = drinks_data.describe()

summary_stats

beer_servings  spirit_servings  wine_servings  \
count     193.000000       193.000000     193.000000   
mean      106.160622        80.994819      49.450777   
std       101.143103        88.284312      79.697598   
min         0.000000         0.000000       0.000000   
25%        20.000000         4.000000       1.000000   
50%        76.000000        56.000000       8.000000   
75%       188.000000       128.000000      59.000000   
max       376.000000       438.000000     370.000000   

       total_litres_of_pure_alcohol  
count                    193.000000  
mean                       4.717098  
std                        3.773298  
min                        0.000000  
25%                        1.300000  
50%                        4.200000  
75%                        7.200000  
max                       14.400000

This takes away the non numeric columns, so lets include them all:

In [21]:
summary = drinks_data.describe(include='all')

summary

country  beer_servings  spirit_servings  wine_servings  \
count           193     193.000000       193.000000     193.000000   
unique          193            NaN              NaN            NaN   
top     Afghanistan            NaN              NaN            NaN   
freq              1            NaN              NaN            NaN   
mean            NaN     106.160622        80.994819      49.450777   
std             NaN     101.143103        88.284312      79.697598   
min             NaN       0.000000         0.000000       0.000000   
25%             NaN      20.000000         4.000000       1.000000   
50%             NaN      76.000000        56.000000       8.000000   
75%             NaN     188.000000       128.000000      59.000000   
max             NaN     376.000000       438.000000     370.000000   

        total_litres_of_pure_alcohol continent  
count                     193.000000       170  
unique                           NaN         5  
top                              NaN        AF  
freq                             NaN        53  
mean                        4.717098       NaN  
std                         3.773298       NaN  
min                         0.000000       NaN  
25%                         1.300000       NaN  
50%                         4.200000       NaN  
75%                         7.200000       NaN  
max                        14.400000       NaN

Here's a summary of the dataset's properties:

- The data types for each column are appropriate, with **beer_servings**, **spirit_servings**, and **wine_servings** being int, and **total_litres_of_pure_alcohol** being a float-point number. **country** and **continent** being objects (which typically denotes strings in pandas).
- There are 23 missing values found in the **continent** column.
- There are 193 records in total - only 170 shown in the continent column.
- On average beer was served the most in the countries with the highest average servings of approximately 106.

Next steps in the EDA will include:

Cleaning the NaNs / Missing values

### **Findings which rows have a NaN:**

### **Indexing:**

In [25]:
# Find all columns that contain a Missing Value (NaN)
drinks_data[drinks_data.isna()['continent']]

country  beer_servings  spirit_servings  \
5               Antigua & Barbuda            102              128   
11                        Bahamas            122              176   
14                       Barbados            143              173   
17                         Belize            263              114   
32                         Canada            240              122   
41                     Costa Rica            149               87   
43                           Cuba             93              137   
50                       Dominica             52              286   
51             Dominican Republic            193              147   
54                    El Salvador             52               69   
68                        Grenada            199              438   
69                      Guatemala             53               69   
73                          Haiti              1              326   
74                       Honduras             69               98   
84                        Jamaica             82               97   
109                        Mexico            238               68   
122                     Nicaragua             78              118   
130                        Panama            285              104   
143             St. Kitts & Nevis            194              205   
144                     St. Lucia            171              315   
145  St. Vincent & the Grenadines            120              221   
174             Trinidad & Tobago            197              156   
184                           USA            249              158   

     wine_servings  total_litres_of_pure_alcohol continent  
5               45                           4.9       NaN  
11              51                           6.3       NaN  
14              36                           6.3       NaN  
17               8                           6.8       NaN  
32             100                           8.2       NaN  
41              11                           4.4       NaN  
43               5                           4.2       NaN  
50              26                           6.6       NaN  
51               9                           6.2       NaN  
54               2                           2.2       NaN  
68              28                          11.9       NaN  
69               2                           2.2       NaN  
73               1                           5.9       NaN  
74               2                           3.0       NaN  
84               9                           3.4       NaN  
109              5                           5.5       NaN  
122              1                           3.5       NaN  
130             18                           7.2       NaN  
143             32                           7.7       NaN  
144             71                          10.1       NaN  
145             11                           6.3       NaN  
174              7                           6.4       NaN  
184             84                           8.7       NaN

### **Dealing with the Missing Data 1: Dropping NaNs:**

In [27]:
drinks_data.dropna()

country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan              0                0              0   
1        Albania             89              132             54   
2        Algeria             25                0             14   
3        Andorra            245              138            312   
4         Angola            217               57             45   
..           ...            ...              ...            ...   
188    Venezuela            333              100              3   
189      Vietnam            111                2              1   
190        Yemen              6                0              0   
191       Zambia             32               19              4   
192     Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol continent  
0                             0.0        AS  
1                             4.9        EU  
2                             0.7        AF  
3                            12.4        EU  
4                             5.9        AF  
..                            ...       ...  
188                           7.7        SA  
189                           2.0        AS  
190                           0.1        AS  
191                           2.5        AF  
192                           4.7        AF  

[170 rows x 6 columns]

Now that you have deleted the 23 rows with NaNs, you will obtain 170 full rows.

The method `.dropna()` by default doesn't do changes inplace, to do so you need to pass `inplace=True`, or set the output to another variable.

In [30]:
drinks_data

country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan              0                0              0   
1        Albania             89              132             54   
2        Algeria             25                0             14   
3        Andorra            245              138            312   
4         Angola            217               57             45   
..           ...            ...              ...            ...   
188    Venezuela            333              100              3   
189      Vietnam            111                2              1   
190        Yemen              6                0              0   
191       Zambia             32               19              4   
192     Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol continent  
0                             0.0        AS  
1                             4.9        EU  
2                             0.7        AF  
3                            12.4        EU  
4                             5.9        AF  
..                            ...       ...  
188                           7.7        SA  
189                           2.0        AS  
190                           0.1        AS  
191                           2.5        AF  
192                           4.7        AF  

[193 rows x 6 columns]

Making a new variable / copy of dataset with the deletion of columns with NAs, so we keep original dataset wihtout the permenant changes

In [32]:
# Make a new variable for the new dataframe which will have the removal of the missing values
df_drinks = drinks_data.copy()
df_drinks.dropna(inplace=True)
df_drinks

country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan              0                0              0   
1        Albania             89              132             54   
2        Algeria             25                0             14   
3        Andorra            245              138            312   
4         Angola            217               57             45   
..           ...            ...              ...            ...   
188    Venezuela            333              100              3   
189      Vietnam            111                2              1   
190        Yemen              6                0              0   
191       Zambia             32               19              4   
192     Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol continent  
0                             0.0        AS  
1                             4.9        EU  
2                             0.7        AF  
3                            12.4        EU  
4                             5.9        AF  
..                            ...       ...  
188                           7.7        SA  
189                           2.0        AS  
190                           0.1        AS  
191                           2.5        AF  
192                           4.7        AF  

[170 rows x 6 columns]

### **Dropping rows: More than one NaN**

Sometimes, you want to drop the row only if more than `N` columns have `NaNs`:

In [35]:
# Drop the rows that contain more than one NaN.
drinks_data.dropna(thresh=len(drinks_data.columns)-1)   # We allow 1 NaN

country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan              0                0              0   
1        Albania             89              132             54   
2        Algeria             25                0             14   
3        Andorra            245              138            312   
4         Angola            217               57             45   
..           ...            ...              ...            ...   
188    Venezuela            333              100              3   
189      Vietnam            111                2              1   
190        Yemen              6                0              0   
191       Zambia             32               19              4   
192     Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol continent  
0                             0.0        AS  
1                             4.9        EU  
2                             0.7        AF  
3                            12.4        EU  
4                             5.9        AF  
..                            ...       ...  
188                           7.7        SA  
189                           2.0        AS  
190                           0.1        AS  
191                           2.5        AF  
192                           4.7        AF  

[193 rows x 6 columns]

If the row contains only 1 NaN, then the row will not be dropped - Indicating here that the NaNs are only found in the continent column.

In [37]:
drinks_data.dropna(thresh=len(drinks_data.columns))   # Same as dropping rows with any NaN

country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan              0                0              0   
1        Albania             89              132             54   
2        Algeria             25                0             14   
3        Andorra            245              138            312   
4         Angola            217               57             45   
..           ...            ...              ...            ...   
188    Venezuela            333              100              3   
189      Vietnam            111                2              1   
190        Yemen              6                0              0   
191       Zambia             32               19              4   
192     Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol continent  
0                             0.0        AS  
1                             4.9        EU  
2                             0.7        AF  
3                            12.4        EU  
4                             5.9        AF  
..                            ...       ...  
188                           7.7        SA  
189                           2.0        AS  
190                           0.1        AS  
191                           2.5        AF  
192                           4.7        AF  

[170 rows x 6 columns]

**We can also drop rows based on NaN on only some columns:**

In [39]:
# Dropping rows with NaN values in these columns
drinks_data.dropna(subset=['beer_servings', 'spirit_servings', 'wine_servings'])

country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan              0                0              0   
1        Albania             89              132             54   
2        Algeria             25                0             14   
3        Andorra            245              138            312   
4         Angola            217               57             45   
..           ...            ...              ...            ...   
188    Venezuela            333              100              3   
189      Vietnam            111                2              1   
190        Yemen              6                0              0   
191       Zambia             32               19              4   
192     Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol continent  
0                             0.0        AS  
1                             4.9        EU  
2                             0.7        AF  
3                            12.4        EU  
4                             5.9        AF  
..                            ...       ...  
188                           7.7        SA  
189                           2.0        AS  
190                           0.1        AS  
191                           2.5        AF  
192                           4.7        AF  

[193 rows x 6 columns]

Nothing was dropped because the columns do not contain any NaN values.

### **Dropping Columns with NaN:**

In [42]:
# Dropping columns that contain NaN
drinks_data.dropna(axis='columns')

country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan              0                0              0   
1        Albania             89              132             54   
2        Algeria             25                0             14   
3        Andorra            245              138            312   
4         Angola            217               57             45   
..           ...            ...              ...            ...   
188    Venezuela            333              100              3   
189      Vietnam            111                2              1   
190        Yemen              6                0              0   
191       Zambia             32               19              4   
192     Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol  
0                             0.0  
1                             4.9  
2                             0.7  
3                            12.4  
4                             5.9  
..                            ...  
188                           7.7  
189                           2.0  
190                           0.1  
191                           2.5  
192                           4.7  

[193 rows x 5 columns]

The whole continent column was dropped as it was the only column which contained NaN values.

## **Dealing with Missing Data 2: Using Fillers to replace NaNs**

### Filling with a simple value

The simplest way to fill a missing value is to fill it with a **constant value**. Also this is the most dangerous way since you have to be very sure of what you input.
- The three main ways to fill the NaN: fill with the **constant value**, do what is known as a **forward and backward fill**, or **mean fill** (average)

In [47]:
# The issue is that the North America values were captured as Not a Number.
drinks_data.fillna('NA', inplace=True)
drinks_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 6 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   country                       193 non-null    object 
 1   beer_servings                 193 non-null    int64  
 2   spirit_servings               193 non-null    int64  
 3   wine_servings                 193 non-null    int64  
 4   total_litres_of_pure_alcohol  193 non-null    float64
 5   continent                     193 non-null    object 
dtypes: float64(1), int64(3), object(2)
memory usage: 9.2+ KB


In [48]:
# Indexing the rows with NA in the continent column
drinks_data[drinks_data['continent']== 'NA']

country  beer_servings  spirit_servings  \
5               Antigua & Barbuda            102              128   
11                        Bahamas            122              176   
14                       Barbados            143              173   
17                         Belize            263              114   
32                         Canada            240              122   
41                     Costa Rica            149               87   
43                           Cuba             93              137   
50                       Dominica             52              286   
51             Dominican Republic            193              147   
54                    El Salvador             52               69   
68                        Grenada            199              438   
69                      Guatemala             53               69   
73                          Haiti              1              326   
74                       Honduras             69               98   
84                        Jamaica             82               97   
109                        Mexico            238               68   
122                     Nicaragua             78              118   
130                        Panama            285              104   
143             St. Kitts & Nevis            194              205   
144                     St. Lucia            171              315   
145  St. Vincent & the Grenadines            120              221   
174             Trinidad & Tobago            197              156   
184                           USA            249              158   

     wine_servings  total_litres_of_pure_alcohol continent  
5               45                           4.9        NA  
11              51                           6.3        NA  
14              36                           6.3        NA  
17               8                           6.8        NA  
32             100                           8.2        NA  
41              11                           4.4        NA  
43               5                           4.2        NA  
50              26                           6.6        NA  
51               9                           6.2        NA  
54               2                           2.2        NA  
68              28                          11.9        NA  
69               2                           2.2        NA  
73               1                           5.9        NA  
74               2                           3.0        NA  
84               9                           3.4        NA  
109              5                           5.5        NA  
122              1                           3.5        NA  
130             18                           7.2        NA  
143             32                           7.7        NA  
144             71                          10.1        NA  
145             11                           6.3        NA  
174              7                           6.4        NA  
184             84                           8.7        NA

All the NaN's in the continent column were changed to NA (North America), so now continent has 193 rows full data.

## **Filling NaN with Mean imputation:**

Loading new dataset into python using pandas library

In [52]:
drinks_mixed = pd.read_csv('drinks_mixed.csv')

drinks_mixed.head()

Unnamed: 0      country  beer_servings  spirit_servings  wine_servings  \
0           0  Afghanistan            0.0              0.0            0.0   
1           1      Albania           89.0            132.0            NaN   
2           2      Algeria           25.0              NaN           14.0   
3           3      Andorra          245.0            138.0          312.0   
4           4       Angola          217.0              NaN           45.0   

   total_litres_of_pure_alcohol continent  
0                           NaN        AS  
1                           4.9        EU  
2                           0.7        AF  
3                           NaN        EU  
4                           5.9        AF

In [53]:
# Checking which columns contain missing values
drinks_mixed.isna().sum()

Unnamed: 0                       0
country                          0
beer_servings                   48
spirit_servings                 51
wine_servings                   45
total_litres_of_pure_alcohol    49
continent                       23
dtype: int64

In [54]:
# Checks datatype and columns that contain missing values
drinks_mixed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 7 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Unnamed: 0                    193 non-null    int64  
 1   country                       193 non-null    object 
 2   beer_servings                 145 non-null    float64
 3   spirit_servings               142 non-null    float64
 4   wine_servings                 148 non-null    float64
 5   total_litres_of_pure_alcohol  144 non-null    float64
 6   continent                     170 non-null    object 
dtypes: float64(4), int64(1), object(2)
memory usage: 10.7+ KB


Lets try to drop all the rows with missing values

In [56]:
# Drop all rows with NaNs
drinks_mixed.dropna()

Empty DataFrame
Columns: [Unnamed: 0, country, beer_servings, spirit_servings, wine_servings, total_litres_of_pure_alcohol, continent]
Index: []

We get nothing, meaning every row in the data has at least one NaN. We can fill with the mean value:

In [58]:
# Filling all the numerical columns with NaN using mean imputation

drinks_mixed.fillna(drinks_mixed.select_dtypes(include='number').mean(), inplace=True)

In [59]:
drinks_mixed

Unnamed: 0      country  beer_servings  spirit_servings  wine_servings  \
0             0  Afghanistan       0.000000         0.000000       0.000000   
1             1      Albania      89.000000       132.000000      49.587838   
2             2      Algeria      25.000000        84.380282      14.000000   
3             3      Andorra     245.000000       138.000000     312.000000   
4             4       Angola     217.000000        84.380282      45.000000   
..          ...          ...            ...              ...            ...   
188         188    Venezuela     100.227586       100.000000       3.000000   
189         189      Vietnam     111.000000        84.380282       1.000000   
190         190        Yemen       6.000000         0.000000      49.587838   
191         191       Zambia     100.227586        19.000000       4.000000   
192         192     Zimbabwe      64.000000        84.380282       4.000000   

     total_litres_of_pure_alcohol continent  
0                        4.673611        AS  
1                        4.900000        EU  
2                        0.700000        AF  
3                        4.673611        EU  
4                        5.900000        AF  
..                            ...       ...  
188                      7.700000        SA  
189                      2.000000        AS  
190                      0.100000        AS  
191                      2.500000        AF  
192                      4.700000        AF  

[193 rows x 7 columns]

In [60]:
drinks_mixed.isna().sum()

Unnamed: 0                       0
country                          0
beer_servings                    0
spirit_servings                  0
wine_servings                    0
total_litres_of_pure_alcohol     0
continent                       23
dtype: int64

## **Handling Duplicated Values in Data**

In [62]:
drinks_duplicated = pd.read_csv('drinks_duplicated.csv')

drinks_duplicated

country  beer_servings  spirit_servings  wine_servings  \
0         Uganda             45                9              0   
1           Iraq              9                3              0   
2      Macedonia            106               27             86   
3      Venezuela            333              100              3   
4     Bangladesh              0                0              0   
..           ...            ...              ...            ...   
334       Sweden            152               60            186   
335  South Korea            140               16              9   
336  Philippines             71              186              1   
337       Mexico            238               68              5   
338         Chad             15                1              1   

     total_litres_of_pure_alcohol continent  
0                             8.3        AF  
1                             0.2        AS  
2                             3.9        EU  
3                             7.7        SA  
4                             0.0        AS  
..                            ...       ...  
334                           7.2        EU  
335                           9.8        AS  
336                           4.6        AS  
337                           5.5       NaN  
338                           0.4        AF  

[339 rows x 6 columns]

In [63]:
# Remove any leading or trailing whitespace from column names
drinks_duplicated.columns = drinks_duplicated.columns.str.strip()

In [64]:
drinks_duplicated.columns

Index(['country', 'beer_servings', 'spirit_servings', 'wine_servings',
       'total_litres_of_pure_alcohol', 'continent'],
      dtype='object')

## Finding Duplicates and Dealing with them

A good way to find duplicated data is to index some columns. Let's index by country, which should be unique in this dataset.

In [67]:
# Indexing the country column and calculating how many times the unique values appear
drinks_duplicated.set_index('country', inplace=True)

In [68]:
drinks_duplicated.index.value_counts()

country
Myanmar      5
Tanzania     5
Suriname     5
Togo         5
Spain        5
            ..
Guatemala    1
Kuwait       1
Yemen        1
Italy        1
Chad         1
Name: count, Length: 193, dtype: int64

We can see that some countries have multiple rows. 

In [70]:
# Resetting the index 
drinks_duplicated.reset_index(inplace=True)

In [71]:
drinks_duplicated.duplicated()  # The first occurance is set to normal, the rest as duplicated

0      False
1      False
2      False
3      False
4      False
       ...  
334    False
335    False
336     True
337    False
338    False
Length: 339, dtype: bool

In [72]:
drinks_duplicated[~drinks_duplicated.duplicated()]

country  beer_servings  spirit_servings  wine_servings  \
0         Uganda             45                9              0   
1           Iraq              9                3              0   
2      Macedonia            106               27             86   
3      Venezuela            333              100              3   
4     Bangladesh              0                0              0   
..           ...            ...              ...            ...   
332       Guyana             93              302              1   
334       Sweden            152               60            186   
335  South Korea            140               16              9   
337       Mexico            238               68              5   
338         Chad             15                1              1   

     total_litres_of_pure_alcohol continent  
0                             8.3        AF  
1                             0.2        AS  
2                             3.9        EU  
3                             7.7        SA  
4                             0.0        AS  
..                            ...       ...  
332                           7.1        SA  
334                           7.2        EU  
335                           9.8        AS  
337                           5.5       NaN  
338                           0.4        AF  

[193 rows x 6 columns]

- This gets rid of duplicates in the dataset, if it already found one record, then it will filter out the ones with False.
- However, this method does not drop rows that are not exactly duplicates, but maybe they are in some key columns that you care about.

## **Handling Invalid Values in our Data**

### Using functions with apply to detect invalid rows and fix them:

In [76]:
# Loading dataset
invalid_measurements = pd.read_csv('measurements_invalid.csv')
invalid_measurements

sensor                        date measurement
0     temperature  2023-04-25 13:53:31.759460          40
1        humidity  2023-04-25 13:53:31.759470          26
2     temperature  2023-04-25 13:53:31.759473         196
3     temperature  2023-04-25 13:53:31.759475           2
4     temperature  2023-04-25 13:53:31.759477          10
...           ...                         ...         ...
9995       weight  2023-04-25 13:53:31.780708          60
9996  temperature  2023-04-25 13:53:31.780710         134
9997       weight  2023-04-25 13:53:31.780712          33
9998     humidity  2023-04-25 13:53:31.780713          77
9999     humidity  2023-04-25 13:53:31.780715          26

[10000 rows x 3 columns]

### Detecting invalid data:

### 1. Filtering the invalid data out

Dealing with invalid data is one of the most difficult parts of data wrangling. You need to understand if that error is something tractable or not. One easy way to detect it is to see if you can run the mean funtion on each column.

In [80]:
invalid_measurements.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sensor       10000 non-null  object
 1   date         10000 non-null  object
 2   measurement  10000 non-null  object
dtypes: object(3)
memory usage: 234.5+ KB


The trick to remove invalid data, or detect it, is to use the `to_numeric` method, which will try to convert supposedly numeric fields into numertic types. The second trick is to set errors to `coerce` such that failures get converted to NaNs; and we know how to handles those.

In [82]:
# Convert each invalid value which cant be converted into a numberical form into Not a Number (NaN)
pd.to_numeric(invalid_measurements['measurement'], errors='coerce')

0        40.0
1        26.0
2       196.0
3         2.0
4        10.0
        ...  
9995     60.0
9996    134.0
9997     33.0
9998     77.0
9999     26.0
Name: measurement, Length: 10000, dtype: float64

We get back our series of numbers with NaNs 

In [84]:
# Run isna() to find all the missing values NaNs
sum(pd.to_numeric(invalid_measurements['measurement'], errors='coerce').isna())

935

We can see we have 935 troublesome values. Let's remove them.

In [86]:
# Convert to numeric - so we get back an index and NaNs - isna() is then applied so NaNs will be a True and values will be False
invalid_measurements_filtered = invalid_measurements[~pd.to_numeric(invalid_measurements['measurement'], errors='coerce').isna()]

Now we can convert those measurements to numeric!

In [88]:
# Convert measurement column to numeric
invalid_measurements_filtered['measurement'] = invalid_measurements_filtered.measurement.astype('int32')

C:\Users\clive\AppData\Local\Temp\ipykernel_15224\2251879846.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  invalid_measurements_filtered['measurement'] = invalid_measurements_filtered.measurement.astype('int32')


In [89]:
# Check the datatypes
invalid_measurements_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9065 entries, 0 to 9999
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sensor       9065 non-null   object
 1   date         9065 non-null   object
 2   measurement  9065 non-null   int32 
dtypes: int32(1), object(2)
memory usage: 247.9+ KB


In [90]:
# Check the summary statistics
invalid_measurements_filtered.describe()

measurement
count  9065.000000
mean    787.018974
std    1507.348461
min     -10.000000
25%      48.000000
50%      97.000000
75%     193.000000
max    5991.000000

We get all the descriptive statistics without errors

### 2. Replacing the invalid Data

In [93]:
# Loading drinks data
drinks = pd.read_csv('drinks.csv')
drinks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 6 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   country                       193 non-null    object 
 1   beer_servings                 193 non-null    int64  
 2   spirit_servings               193 non-null    int64  
 3   wine_servings                 193 non-null    int64  
 4   total_litres_of_pure_alcohol  193 non-null    float64
 5   continent                     170 non-null    object 
dtypes: float64(1), int64(3), object(2)
memory usage: 9.2+ KB


In [94]:
drinks[drinks['continent'].isna()]

country  beer_servings  spirit_servings  \
5               Antigua & Barbuda            102              128   
11                        Bahamas            122              176   
14                       Barbados            143              173   
17                         Belize            263              114   
32                         Canada            240              122   
41                     Costa Rica            149               87   
43                           Cuba             93              137   
50                       Dominica             52              286   
51             Dominican Republic            193              147   
54                    El Salvador             52               69   
68                        Grenada            199              438   
69                      Guatemala             53               69   
73                          Haiti              1              326   
74                       Honduras             69               98   
84                        Jamaica             82               97   
109                        Mexico            238               68   
122                     Nicaragua             78              118   
130                        Panama            285              104   
143             St. Kitts & Nevis            194              205   
144                     St. Lucia            171              315   
145  St. Vincent & the Grenadines            120              221   
174             Trinidad & Tobago            197              156   
184                           USA            249              158   

     wine_servings  total_litres_of_pure_alcohol continent  
5               45                           4.9       NaN  
11              51                           6.3       NaN  
14              36                           6.3       NaN  
17               8                           6.8       NaN  
32             100                           8.2       NaN  
41              11                           4.4       NaN  
43               5                           4.2       NaN  
50              26                           6.6       NaN  
51               9                           6.2       NaN  
54               2                           2.2       NaN  
68              28                          11.9       NaN  
69               2                           2.2       NaN  
73               1                           5.9       NaN  
74               2                           3.0       NaN  
84               9                           3.4       NaN  
109              5                           5.5       NaN  
122              1                           3.5       NaN  
130             18                           7.2       NaN  
143             32                           7.7       NaN  
144             71                          10.1       NaN  
145             11                           6.3       NaN  
174              7                           6.4       NaN  
184             84                           8.7       NaN

In [141]:
# Checking last 20 lines of original dataset
with open("drinks.csv", "r") as file:
    lines = file.readlines()[-20:]  # Get the last 20 lines
    for line in lines:
        print(line.strip())

Tonga,36,21,5,1.1,OC
Trinidad & Tobago,197,156,7,6.4,NA
Tunisia,51,3,20,1.3,AF
Turkey,51,22,7,1.4,AS
Turkmenistan,19,71,32,2.2,AS
Tuvalu,6,41,9,1.0,OC
Uganda,45,9,0,8.3,AF
Ukraine,206,237,45,8.9,EU
United Arab Emirates,16,135,5,2.8,AS
United Kingdom,219,126,195,10.4,EU
Tanzania,36,6,1,5.7,AF
USA,249,158,84,8.7,NA
Uruguay,115,35,220,6.6,SA
Uzbekistan,25,101,8,2.4,AS
Vanuatu,21,18,11,0.9,OC
Venezuela,333,100,3,7.7,SA
Vietnam,111,2,1,2.0,AS
Yemen,6,0,0,0.1,AS
Zambia,32,19,4,2.5,AF
Zimbabwe,64,18,4,4.7,AF


NA is in Trinidad & Tabago and USA.

#### Replacing:

In [147]:
# Replace the NaNs with NA (North America)
drinks.continent.replace(np.nan, 'NA', inplace=True)

C:\Users\clive\AppData\Local\Temp\ipykernel_15224\3762707118.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  drinks.continent.replace(np.nan, 'NA', inplace=True)


In [149]:
drinks

country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan              0                0              0   
1        Albania             89              132             54   
2        Algeria             25                0             14   
3        Andorra            245              138            312   
4         Angola            217               57             45   
..           ...            ...              ...            ...   
188    Venezuela            333              100              3   
189      Vietnam            111                2              1   
190        Yemen              6                0              0   
191       Zambia             32               19              4   
192     Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol continent  
0                             0.0        AS  
1                             4.9        EU  
2                             0.7        AF  
3                            12.4        EU  
4                             5.9        AF  
..                            ...       ...  
188                           7.7        SA  
189                           2.0        AS  
190                           0.1        AS  
191                           2.5        AF  
192                           4.7        AF  

[193 rows x 6 columns]

In [151]:
drinks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 6 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   country                       193 non-null    object 
 1   beer_servings                 193 non-null    int64  
 2   spirit_servings               193 non-null    int64  
 3   wine_servings                 193 non-null    int64  
 4   total_litres_of_pure_alcohol  193 non-null    float64
 5   continent                     193 non-null    object 
dtypes: float64(1), int64(3), object(2)
memory usage: 9.2+ KB


In [153]:
The columns have been changed
drinks.continent.value_counts()

continent
AF    53
EU    45
AS    44
NA    23
OC    16
SA    12
Name: count, dtype: int64

## There is a complex way that is a little but more convoluted:

In [169]:
drinks = pd.read_csv('drinks.csv', keep_default_na=False) # Disables orginal list os NaNs - takes whatever values you pass
drinks

country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan              0                0              0   
1        Albania             89              132             54   
2        Algeria             25                0             14   
3        Andorra            245              138            312   
4         Angola            217               57             45   
..           ...            ...              ...            ...   
188    Venezuela            333              100              3   
189      Vietnam            111                2              1   
190        Yemen              6                0              0   
191       Zambia             32               19              4   
192     Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol continent  
0                             0.0        AS  
1                             4.9        EU  
2                             0.7        AF  
3                            12.4        EU  
4                             5.9        AF  
..                            ...       ...  
188                           7.7        SA  
189                           2.0        AS  
190                           0.1        AS  
191                           2.5        AF  
192                           4.7        AF  

[193 rows x 6 columns]

In [171]:
drinks.continent.value_counts()

continent
AF    53
EU    45
AS    44
NA    23
OC    16
SA    12
Name: count, dtype: int64

## Tip: Always check the values in the original file to ensure that it matches the na_values